### PR TITLE
Adding Layer Ablation

### PR DESCRIPTION
**Pull request by vivekmig**
This PR adds layer ablation, which computes the attribution of the input / output of a layer, by setting each neuron or group of neurons to a given baseline and measuring the difference in output value. This is analogous to feature ablation for layer attribution, with baselines and feature masks given with respect to the layer input / output size.

Differential Revision: D18670637

### CODE REVIEW

**Comment by NarineK** [comment]
Let's add a brief description to this one and Neuron Ablation as well.

File: captum/attr/_core/layer/layer_ablation.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+            >>> # group (0 - 2) per example are the same.
+            >>> # The attributions can be calculated as follows:
+            >>> # layer mask has dimensions 1 x 3 x 3
+            >>> layer_mask = torch.tensor([[[0,0,0],[1,1,1],
+            >>>                             [2,2,2]]])
+            >>> attr = ablator.attribute(input, target=1,
+            >>>                          layer_mask=layer_mask)
+        """
+
+        def layer_forward_func(*args):
```
Comment (NarineK) [inline captum/attr/_core/layer/layer_ablation.py]: nit: could this be passed in 2 arguments, like ?
```
layer_input =inputs
original_inputs = additional_forward_args
```

File: captum/attr/_core/layer/layer_ablation.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+            if (
+                device_ids is None
+                and isinstance(self.forward_func, torch.nn.DataParallel)
+                and self.forward_func.device_ids is not None
+            ):
+                device_ids = self.forward_func.device_ids
+
+            all_layer_inputs = {}
+            if device_ids is not None:
+                scattered_layer_input, _ = scatter_kwargs(
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/layer/layer_ablation.py]: Good point, will update!

File: captum/attr/_core/layer/layer_ablation.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+            layer_input = args[0]
+            original_inputs = args[1:]
+
+            device_ids = self.device_ids
+            if (
+                device_ids is None
+                and isinstance(self.forward_func, torch.nn.DataParallel)
+                and self.forward_func.device_ids is not None
+            ):
+                device_ids = self.forward_func.device_ids
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/layer/layer_ablation.py]: Good point, the getattr syntax is shorter! I think it might be good to still leave the DataParallel instance check though, it's possible a model might have that attribute and not be DataParallel / not need to be scattered?

File: captum/attr/_core/layer/layer_ablation.py (hunk 1 above)
Comment (vivekmig) [reply captum/attr/_core/layer/layer_ablation.py]: This would work if layer_input is guaranteed to be 1 tensor / element as it is now, but once that becomes variable, I'm not sure how to cleanly separate it, since _run_forward in feature ablation will concatenate the inputs and additional_args tuples?

File: captum/attr/_core/layer/layer_ablation.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+            (inputs + additional_forward_args)
+            if additional_forward_args is not None
+            else inputs
+        )
+
+        ablator = FeatureAblation(layer_forward_func)
+
+        return ablator.attribute(
+            layer_eval,
+            baselines=baselines,
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/layer/layer_ablation.py]: Makes sense, yeah, it's a different here from baselines in other algorithms. I think since we want to evaluate essentially "turning off" each layer input / output, it's not clear what the layer values would be for a given input baseline. Maybe we can document the difference between this baseline and others more clearly? Or rename the argument to something like layer_baseline instead?

**Comment by vivekmig** [comment]
Thanks for the review! Addressed most of the comments.
1. I've renamed to LayerFeatureAblation and also refer to ablating groups of layer input / output elements. I think the confusion in naming is based on defining each neuron as a unique set of weights as opposed to each scalar in the layer as an input / output "neuron", so hopefully this is clearer!

2. Makes sense, yes, this isn't completely consistent with baselines for Layer IG and Conductance, but I do think it's preferable to keep it this way here, since we want to essentially "turn off" each layer input / output neuron, and that's more difficult with an input baseline. We could rename the baselines argument to layer_baselines to make the distinction clearer, what do you think?

File: captum/attr/_core/layer/layer_feature_ablation.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+            all_layer_inputs = {}
+            if device_ids is not None:
+                scattered_layer_input = scatter(layer_input, target_gpus=device_ids)
+                for tensor in scattered_layer_input:
+                    all_layer_inputs[tensor.device] = tensor
+            else:
+                all_layer_inputs[layer_input.device] = layer_input
+
+            def forward_hook(module, inp, out=None):
+                if inp[0].device not in all_layer_inputs:
```
Comment (NarineK) [inline captum/attr/_core/layer/layer_feature_ablation.py]: I moved `_extract_device` to common - we could use it here too ? https://github.com/pytorch/captum/pull/222/files#diff-3ff60ae8a182cf817df47a82964d8730R438

File: captum/attr/_core/layer/layer_feature_ablation.py [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+
+        def layer_forward_func(*args):
+            layer_input = args[0]
+            original_inputs = args[1:]
+
+            device_ids = self.device_ids
+            if device_ids is None and isinstance(
+                self.forward_func, torch.nn.DataParallel
+            ):
+                device_ids = getattr(self.forward_func, "device_ids", None)
```
Comment (NarineK) [inline captum/attr/_core/layer/layer_feature_ablation.py]: I'm not able to comment in our previous conversation thread about the `device_ids ` - I'll comment here:
I think currently we have some places where we check with `torch.nn.DataParallel` and some cases where we don't (e.g in `gradient.py`).
If we don't explicitly check for `torch.nn.DataParallel` then it is more generic but the user might not always want it as you mentioned. 
But from other point of view they can always set `self.forward_func.device_ids=None` if they don't want the computations to be performed in a distributed manner ?

File: captum/attr/_core/layer/layer_feature_ablation.py (hunk 5 above)
Comment (vivekmig) [reply captum/attr/_core/layer/layer_feature_ablation.py]: Sounds good, I will wait for that PR to land and then update

File: captum/attr/_core/layer/layer_feature_ablation.py (hunk 6 above)
Comment (vivekmig) [reply captum/attr/_core/layer/layer_feature_ablation.py]: Makes sense, sounds good, it would be good to choose one approach and stay consistent. Should we not check for torch.nn.DataParallel everywhere when accessing device_ids and update documentation to reflect this?

Comment (NarineK) [reply captum/attr/_core/layer/layer_feature_ablation.py]: Yeah, probably that would be good - in case we want to also support ModelParallel, DistributedDataParallel in the future and custom forward functions.

Comment (vivekmig) [reply captum/attr/_core/layer/layer_feature_ablation.py]: Sounds good
